compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -336,13 +337,30 @@ impl<'a> Linker for GccLinker<'a> {
         }
     }
 
-    fn link_dylib(&mut self, lib: Symbol) {
+    fn link_dylib(&mut self, lib: Symbol, verbatim: bool, as_needed: bool) {
+        if !as_needed {
+            if self.sess.target.is_like_osx {
+                // FIXME(81490): ld64 doesn't support these flags but macOS 11
+                // has -needed-l{} / -needed_library {}
+                // but we have no way to detect that here.
+                self.sess.warn("`as-needed` modifier not implemented yet for ld64");
+            } else {
+                self.linker_arg("--no-as-needed");
```

Comment:
Different targets currently have different defaults with regards to `--as-needed` (you can see that many targets pass `-Wl,--as-needed` to the linker from their target specs).
I want to improve that (with more target options, and less raw linker flags), but right now I think `fn link_dylib` should specify neither `--as-needed` nor `--no-as-needed` unless the linking modifier was passed explicitly.

Reply:
>I want to improve that (with more target options, and less raw linker flags)

Done in https://github.com/rust-lang/rust/pull/83587.
Looks like new target options are not needed yet.

Reply:
The logic for adding `--(no-)as-needed` should look approximately like this:
```rust
if self.sess.target.is_like_osx {
    // FIXME
} else if self.sess.target.linker_is_gnu {
    self.linker_arg("--(no-)as-needed");
} else {
    // Do nothing
}
```
because there are non-`linker_is_gnu` LD-like linkers that don't understand `--(no-)as-needed`, MinGW being the first example.